File: src/main/java/net/gtaun/util/event/EventManagerChild.java
package net.gtaun.util.event;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;

class EventManagerChild implements EventManagerNode
{
	private EventManager rootEventManager;
	private EventManagerNode parentEventManager;
	
	private Collection<HandlerEntry> handlerEntries;
	private Collection<EventManagerNode> childs;
	
	private boolean isDestroyed;
	
	
	EventManagerChild(EventManager root)
	{
		this(root, null);
	}
	
	@Override
	protected void finalize() throws Throwable
	{
		super.finalize();
		if (!isDestroyed) destroy();
	}
	
	private EventManagerChild(EventManager root, EventManagerNode parent)
	{
		rootEventManager = root;
		parentEventManager = parent;
		handlerEntries = new HashSet<>();
		childs = new ArrayList<>();
	}
	
	private void checkDestroyState() throws IllegalStateException
	{
		if (isDestroyed) throw new IllegalStateException("EventManagerNode has been destroyed!");	
	}
	
	private void destroyChild(EventManagerNode node)
	{
		if (isDestroyed) return;
		childs.remove(node);
	}
	
	@Override
	public <E extends Event> HandlerEntry registerHandler(Class<E> type, short priority, Attentions concerns, EventHandler<E> handler)
	{
		checkDestroyState();
		
		final HandlerEntry rootEntry = rootEventManager.registerHandler(type, priority, concerns, handler);
		final HandlerEntry entry = new AbstractHandlerEntry(type, concerns, handler, priority)
		{
			@Override
			public EventManager getEventManager()
			{
				return EventManagerChild.this;
			}
			
			@Override
			public void cancel()
			{
				rootEntry.cancel();
				isCanceled = true;
				if (!isDestroyed) handlerEntries.remove(this);
			}
		};
		
		handlerEntries.add(entry);
		return entry;
	}
	
	@Override
	public <E extends Event> void dispatchEvent(ThrowableHandler handler, E event, Object... objects)
	{
		checkDestroyState();
		rootEventManager.dispatchEvent(handler, event, objects);
	}
	
	@Override
	public EventManagerNode createChildNode()
	{
		EventManagerNode node = rootEventManager.createChildNode();
		childs.add(node);
		return node;
	}
	
	public void cancelAll()
	{
		List<HandlerEntry> entries = new ArrayList<>(handlerEntries);
		for (HandlerEntry entry : entries) entry.cancel();
		handlerEntries.clear();
	}
	
	@Override
	public void destroy()
	{
		if (parentEventManager instanceof EventManagerChild && parentEventManager.isDestroy() == false)
		{
			((EventManagerChild)parentEventManager).destroyChild(this);
		}
		
		isDestroyed = true;
		
		for (EventManagerNode node : childs) node.destroy();
		childs.clear();
		
		for (HandlerEntry entry : handlerEntries) entry.cancel();
		handlerEntries.clear();
	}
	
	@Override
	public boolean isDestroy()
	{
		return isDestroyed;
	}
	
	@Override
	public EventManager getParent()
	{
		return rootEventManager;
	}
}

File: src/main/java/net/gtaun/util/event/EventManagerChild.kt
package net.gtaun.util.event

internal open class EventManagerChild constructor(override val parent: EventManager,
                                                          private val parentEventManager: EventManagerNode? = null,
                                                          override var isDestroyed: Boolean = false) :
        EventManagerNode {

    override var throwableHandler: ThrowableHandler = EventManagerRoot.DEFAULT_THROWABLE_HANDLER

    private val handlerEntries: MutableList<HandlerEntry> = mutableListOf()
    private val children: MutableList<EventManagerNode> = mutableListOf()

    @Throws(Throwable::class)
    protected fun finalize() {
        if (!isDestroyed)
            destroy()
    }

    @Throws(IllegalStateException::class)
    private fun checkDestroyState() {
        if (isDestroyed)
            throw IllegalStateException("EventManagerNode has been destroyed!")
    }

    private fun destroyChild(node: EventManagerNode) {
        if (isDestroyed) return
        children.remove(node)
    }

    override fun <E : Event> registerHandler(type: Class<E>, handler: EventHandler<E>, priority: HandlerPriority,
                                             concerns: Attentions): HandlerEntry {
        checkDestroyState()

        val rootEntry = parent.registerHandler(type, handler, priority, concerns)
        val entry = object : AbstractHandlerEntry(type, concerns, handler, priority.value) {
            override val eventManager: EventManager
                get() = this@EventManagerChild

            override fun cancel() {
                rootEntry.cancel()
                isCanceled = true
                if (!isDestroyed)
                    handlerEntries.remove(this)
            }
        }

        handlerEntries.add(entry)
        return entry
    }

    override fun <E : Event> dispatchEvent(handler: ThrowableHandler?, event: E, vararg objects: Any?) {
        checkDestroyState()
        parent.dispatchEvent(handler, event, *objects)
    }

    override fun createChildNode(): EventManagerNode {
        val node = parent.createChildNode()
        children.add(node)
        return node
    }

    override fun cancelAll() {
        handlerEntries.forEach { it.cancel() }
        handlerEntries.clear()
    }

    override fun destroy() {
        if (parentEventManager is EventManagerChild && parentEventManager.isDestroyed == false) {
            parentEventManager.destroyChild(this)
        }

        isDestroyed = true

        children.forEach { it.destroy() }
        children.clear()

        handlerEntries.forEach { it.cancel() }
        handlerEntries.clear()
    }
}
